packages/angular_devkit/build_webpack/src/app-shell/index.ts

```diff
@@ -0,0 +1,181 @@
+import { BuildWebpackServerSchema } from '../server/schema';
+import { BuildWebpackAppShellSchema } from './schema';
+
+
+export class AppShellBuilder implements Builder<BuildWebpackAppShellSchema> {
+
+  constructor(public context: BuilderContext) { }
+
+  run(builderConfig: BuilderConfiguration<BuildWebpackAppShellSchema>): Observable<BuildEvent> {
+    const options = builderConfig.options;
+
+    return new Observable<BuildEvent>(obs => {
+      let success = true;
+      const subscription = merge(
+        this.build(options.serverTarget, {}),
+        this.build(options.browserTarget, { watch: false }),
+      ).subscribe((event: BuildEvent) => {
+        // TODO: once we support a better build event, add support for merging two event streams
+        // together.
+        success = success && event.success;
+      }, error => {
+        obs.error(error);
+      }, () => {
+        obs.next({ success });
+        obs.complete();
+      });
+
+      // Allow subscriptions to us to unsubscribe from each builds at the same time.
+      return () => subscription.unsubscribe();
+    }).pipe(
+      switchMap(event => {
+        if (!event.success) {
+          return of(event);
+        }
+
+        return this.renderUniversal(options);
+      }),
+    );
+  }
+
+  build(targetString: string, overrides: {}) {
+    const architect = this.context.architect;
+    const [project, target, configuration] = targetString.split(':');
+
+    // Override browser build watch setting.
+    const builderConfig = architect.getBuilderConfiguration<{}>({
+      project,
+      target,
+      configuration,
+      overrides,
+    });
+
+    return architect.getBuilderDescription(builderConfig).pipe(
+      concatMap(description => architect.validateBuilderOptions(builderConfig, description)),
+      switchMap(config => architect.run(config, this.context)),
+    );
+  }
+
+  getServerModuleBundlePath(options: BuildWebpackAppShellSchema) {
+    const architect = this.context.architect;
+
+    return new Observable<Path>(obs => {
+      if (options.renderScript) {
+        obs.next(join(this.context.workspace.root, options.renderScript));
+
+        return obs.complete();
+      } else {
+        const [project, target, configuration] = options.serverTarget.split(':');
+        const builderConfig = architect.getBuilderConfiguration<BuildWebpackServerSchema>({
+          project,
+          target,
+          configuration,
+        });
+
+        return architect.getBuilderDescription(builderConfig).pipe(
+          concatMap(description => architect.validateBuilderOptions(builderConfig, description)),
+          switchMap(config => {
+            const outputPath = join(this.context.workspace.root, config.options.outputPath);
+
+            return this.context.host.list(outputPath).pipe(
+              switchMap(files => {
+                const re = /^main\.(?:[a-zA-Z0-9]{20}\.)?(?:bundle\.)?js$/;
```

Comment:
Not a big fan of this way of determining the mainfile but don't have a better way either. Sounds like something we should take into consideration for build graph information in the future. Even then it sounds complicated because it is semantic information of the output contents.

Reply:
Note this is just for the default.